cmd/clusterctl/pkg/client/repository/repository_github.go

```diff
@@ -0,0 +1,301 @@
+// Repositories must use versioned releases, including the "latest" meta version
+// (https://help.github.com/en/github/administering-a-repository/linking-to-releases#linking-to-the-latest-release).
+type gitHubRepository struct {
+	providerConfig           config.Provider
+	configVariablesClient    config.VariablesClient
+	authenticatingHTTPClient *http.Client
+	owner                    string
+	repository               string
+	defaultVersion           string
+	rootPath                 string
+	componentsPath           string
+	injectClient             *github.Client
+}
+
+var _ Repository = &gitHubRepository{}
+
+// DefaultVersion returns defaultVersion field of gitHubRepository struct
+func (g *gitHubRepository) DefaultVersion() string {
+	return g.defaultVersion
+}
+
+// RootPath returns rootPath field of gitHubRepository struct
+func (g *gitHubRepository) RootPath() string {
+	return g.rootPath
+}
+
+// ComponentsPath returns componentsPath field of gitHubRepository struct
+func (g *gitHubRepository) ComponentsPath() string {
+	return g.componentsPath
+}
+
+// GetFile returns a file for a given provider version
+func (g *gitHubRepository) GetFile(version, path string) ([]byte, error) {
+	release, err := g.getReleaseByTag(version)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get GitHub release %s", version)
+	}
+
+	// download files from the release
+	files, err := g.downloadFilesFromRelease(release, path)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to download files from GitHub release %s", version)
+	}
+
+	return files, nil
+}
+
+// newGitHubRepository returns a gitHubRepository implementation
+func newGitHubRepository(providerConfig config.Provider, configVariablesClient config.VariablesClient) (*gitHubRepository, error) {
+	if configVariablesClient == nil {
+		return nil, errors.New("invalid arguments: configVariablesClient can't be nil")
+	}
+
+	rURL, err := url.Parse(providerConfig.URL())
+	if err != nil {
+		return nil, errors.Wrap(err, "invalid url")
+	}
+
+	// Check if the url is a github repository
+	if !(rURL.Scheme == httpsScheme && rURL.Host == githubDomain) {
+		return nil, errors.New("invalid url: a GitHub repository url should start with https://github.com")
+	}
+
+	// Check if the repository Path is in the expected format.
+	t := strings.Split(strings.TrimPrefix(rURL.Path, "/"), "/")
+	if len(t) < 5 || t[2] != githubReleaseRepository {
+		return nil, errors.Errorf(
+			"invalid url: a GitHub repository url should be in the form https://github.com/{owner}/{Path}/%s/{latest|version-tag}/{componentsClient.yaml}",
+			githubReleaseRepository,
+		)
+	}
+
+	// extracts all the info from path
+	owner := t[0]
+	repository := t[1]
+	defaultVersion := t[3]
+	path := strings.Join(t[4:], "/")
+
+	// use path's directory as a rootPath
+	rootPath := filepath.Dir(path)
+	// use the file name (if any) as componentsPath
+	componentsPath := getComponentsPath(path, rootPath)
```

Comment:
`getComponentsPath` was introduced to address a comment above and to avoid a piece of ugly code; there is also a dedicated unit test now. IMO we should not rollback to inline expressions.

Reply:
https://github.com/kubernetes-sigs/cluster-api/pull/1909#discussion_r358896073
here is the previous comment about this change